src/Type/WebMozartAssert/AssertTypeSpecifyingExtension.php

```diff
@@ -485,31 +486,56 @@ private static function getExpressionResolvers(): array
 						$value2->value
 					);
 				},
-				'greaterThan' => static function (Scope $scope, Arg $value, Arg $limit): Expr {
+				'greaterThan' => static function (Scope $scope, Arg $value, Arg $limit): ?Expr {
+					$valueType = $scope->getType($value->value);
+					if ((new IntegerType())->isSuperTypeOf($valueType)->no()) {
+						return null;
+					}
+
 					return new Greater(
 						$value->value,
 						$limit->value
 					);
 				},
-				'greaterThanEq' => static function (Scope $scope, Arg $value, Arg $limit): Expr {
+				'greaterThanEq' => static function (Scope $scope, Arg $value, Arg $limit): ?Expr {
+					$valueType = $scope->getType($value->value);
+					if ((new IntegerType())->isSuperTypeOf($valueType)->no()) {
+						return null;
+					}
+
 					return new GreaterOrEqual(
 						$value->value,
 						$limit->value
 					);
 				},
-				'lessThan' => static function (Scope $scope, Arg $value, Arg $limit): Expr {
+				'lessThan' => static function (Scope $scope, Arg $value, Arg $limit): ?Expr {
+					$valueType = $scope->getType($value->value);
+					if ((new IntegerType())->isSuperTypeOf($valueType)->no()) {
+						return null;
+					}
+
 					return new Smaller(
 						$value->value,
 						$limit->value
 					);
 				},
-				'lessThanEq' => static function (Scope $scope, Arg $value, Arg $limit): Expr {
+				'lessThanEq' => static function (Scope $scope, Arg $value, Arg $limit): ?Expr {
+					$valueType = $scope->getType($value->value);
+					if ((new IntegerType())->isSuperTypeOf($valueType)->no()) {
+						return null;
+					}
+
 					return new SmallerOrEqual(
 						$value->value,
 						$limit->value
 					);
 				},
-				'range' => static function (Scope $scope, Arg $value, Arg $min, Arg $max): Expr {
+				'range' => static function (Scope $scope, Arg $value, Arg $min, Arg $max): ?Expr {
+					$valueType = $scope->getType($value->value);
+					if ((new IntegerType())->isSuperTypeOf($valueType)->no()) {
```

Comment:
What about `mixed` or `float|int`? Do they demonstrate this bug or not?

Reply:
I assume you're trying to find out if I can use `!(new IntegerType())->isSuperTypeOf($valueType)->yes()` instead? :)
Apparently they are not affected, but I need to be able to use it with `int|null` to not break e.g. `nullOrRange`. If that makes sense. I'm not sure what the best solution for that is

Reply:
Aaah, I understand why. At that point, when the expression is built, it is still the int|null union of course..